sklearn/tree/_criterion.pyx

```diff
@@ -1325,3 +1325,122 @@ cdef class FriedmanMSE(MSE):
+        """Compute a proxy of the impurity reduction.
+
+        This method is used to speed up the search for the best split.
+        It is a proxy quantity such that the split that maximizes this value
+        also maximizes the impurity improvement. It neglects all constant terms
+        of the impurity decrease for a given split.
+
+        The absolute impurity improvement is only computed by the
+        impurity_improvement method once the best split has been found.
+
+        Poisson proxy is:
+            - 1/n * sum(y_i * log(y_pred)) = -mean(y_i) * log(mean(y_i))
+        """
+        cdef SIZE_t k
+        cdef double proxy_impurity_left = 0.0
+        cdef double proxy_impurity_right = 0.0
+        cdef double y_mean_left = 0.
+        cdef double y_mean_right = 0.
+
+        for k in range(self.n_outputs):
+            if (self.sum_left[k] <= 0) or (self.sum_right[k] <= 0):
+                # Poisson loss does not allow non-positive predictions. We
+                # therefore forbid splits that have child nodes with
+                # sum(y_i) <= 0.
+                return -INFINITY
+            else:
+                y_mean_left = self.sum_left[k] / self.weighted_n_left
+                y_mean_right = self.sum_right[k] / self.weighted_n_right
+                proxy_impurity_left += y_mean_left * log(y_mean_left)
+                proxy_impurity_right += y_mean_right * log(y_mean_right)
+
+        return - proxy_impurity_left - proxy_impurity_right
+
+    cdef void children_impurity(self, double* impurity_left,
+                                double* impurity_right) nogil:
+        """Evaluate the impurity in children nodes.
+
+        i.e. the impurity of the left child (samples[start:pos]) and the
+        impurity the right child (samples[pos:end]) for Poisson.
+        """
+        cdef const DOUBLE_t[:, ::1] y = self.y
+
+        cdef SIZE_t start = self.start
+        cdef SIZE_t pos = self.pos
+        cdef SIZE_t end = self.end
+
+        cdef SIZE_t i, p, k
+        cdef DOUBLE_t y_mean = 0.
+        cdef DOUBLE_t w = 1.0
+
+        impurity_left[0] = 0.
+        for k in range(self.n_outputs):
+            y_mean = self.sum_left[k] / self.weighted_n_left
+            for p in range(start, pos):
+                i = self.samples[p]
+
+                if self.sample_weight != NULL:
+                    w = self.sample_weight[i]
+
+                impurity_left[0] += w * xlogy(y[i, k], y[i, k] / y_mean)
+        impurity_left[0] /= self.weighted_n_left * self.n_outputs
+
+        impurity_right[0] = 0.
+        for k in range(self.n_outputs):
+            y_mean = self.sum_right[k] / self.weighted_n_right
+            for p in range(pos, end):
+                i = self.samples[p]
+
+                if self.sample_weight != NULL:
+                    w = self.sample_weight[i]
+
+                impurity_right[0] += w * xlogy(y[i, k], y[i, k] / y_mean)
+        impurity_right[0] /= self.weighted_n_right * self.n_outputs
```

Comment:
this bit of code seems to be duplicated 3 times. Would it make sense to extract it into a private method so that we could use it here and in `node_impurity`?

Reply:
Same could be done for MAE.